### ISSUE TITLE
Build with make - discussion

### ISSUE THREAD

**Issue by stanthomas** [Microsoft/VSLinux#7]
There are two arguments for using `make` (GNU make) on the target Linux system for builing projects in VCLinux : performance and flexibility.

On Windows, the VC++ build process is a sequence of steps where each source file is compiled to an object file and then object files combined into an application (or library) by executing the appropriate CL command for each step under the control of MSBuild. This works well and the ability to run pre- and post-build commands allows for customisation.

VCLinux build processing closely follows that of native Windows applications. However, pre- and post-build commands are missing (1.0.4) and, because it's building on a remote target, the build proceeds slowly.

The VCLinux build steps (1.0.4) are:
1. validate ssh communication with, and architecture of, Linux system
2. copy all project source files to be compiled via ssh to Linux
3. execute g++ command to compile each source file in turn
4. copy object file from each successful compilation back to Windows
5. link objects into executable (or static or shared library)
6. copy executable back to Windows

Personally, I'm quite happy with this process in principle. My applications consists exclusively of C++ source that is built into one or more static libraries, shared libraries and executables. But using `make` seems to offer opportunities to offload some work, both development and execution, to an existing tool.

**Performance**
Because everything has to be done at arms length on a remote Linux system, building is much slower than on Windows; steps are executed sequentially and copying takes time.

It does seem that using `make` on the Linux target and building as a batch process would offer performance improvements:
1. single command via ssh for build
2. parallel compilation capabilities in `make` (`--jobs` option)

An additional perfomance improvement might be to not copy object files back to Windows. The executable is required for debugging with gdb and must be copied. Object files are only to satisfy MSBuild and a locally generated zero-length file with the appropriate name and time stamp would suffice.

**Flexibility**
In addition, basing VCLinux build on `make` would allow users to supply their own makefiles. This would give complete control over the process and work-around issues such as pre- and post-build steps, non-g++ compilation, cross-compilation, etc. There are simply too many variations for VCLinux to cover all the options.

**So how would it work?**
I don't want to have to write a makefile for my projects and would want VCLinux to generate one from the project settings. This is what Eclipse CDT does. Eclipse will create / recreate a makefile on demand from the project settings and contents. Thereafter, Eclipse builds the project using this makefile as it would with a user-supplied one. It's rather clumsy in that the user must regenerate the makefile following any changes to the project, including the addition of new source files.
I'd like to see VCLinux automatically generate its makefile on every build; creating the makefile is quick and build-specific makefiles can be simpler and need only reference the source files affected. It does not need to be a makefile that can be used outside VCLinux.

At the same time it should be possible to give VCLinux the makefile to use. It would have to be part of the Visual Studio project and 'live' on the Windows host. But it would be opaque to VCLinux, i.e. I'm not suggesting it should be parsed by VCLinux. Rules would need to be established for the build targets, e.g. make debug, make clean and what and where the generated executable (or library) wiil be; it being the users responsibility to ensure that everything corresponds (with great power comes great responsibility, eh?).

**Comment by itodirel** [comment]
Thank you for your detailed analysis. Here are some replies.

Regarding the build events, we used to only have local **Pre/Post** build events. And a local **Pre** link event. And we only had one remote **Post** build event. We corrected this symmetry bug, and we now have additionally a remote Pre build event, and a remote Pre link event. This is already done and will be in 1.0.5. If you need more events or more control (Post-Link event?) in this area, please, please let us know!

In 1.0.5, we also fixed a scalability and build/copy performance issues. Specifically, before, the build and source copy was not incremental. Now copy sources is incremental and only copies the files that where changed locally. And compile only compiles the changed files. The incremental compile is not completely done, I call it "incremental build v1" because it does not track the header files incrementally, and if it detects any changes in any headers, it tends to compile everything, but it works fine for the most part, and you will see major performance benefits, and it's much, much faster. In the future, we'll track the includes, and have a more complete and optimized up to date check for compile. We also fixed a scalability issue, where a task, MakeClCommandLine would make connections and call into the remote system for every single file, so if you had a lot of files, it would grow linearly by the number of files. Now it doesn't do that, and completes in milliseconds. So the scalability issue is gone as well.

We have even more opportunities for improving the performance. I've been working (prototype) on a connection cacher, where you can reuse existing connections to the same system. That way, you do not have to make them, and can save significant time. I've been testing it some, and with it on, it's almost as you're executing command on the local machine plus a fixed constant overhead. Plus, the number of commands you execute then also becomes less important. This will not make it in 1.0.5 or 1.0.6, unfortunately. We also want to add in (1.0.6 perhaps) parallel builds, where we can compile the sources in parallel, by a user supplied number of allowed parallel jobs. That would significantly improve the build times. And in the future when we add cross-compile, it will run compile on your local Windows machine, so it will be even faster.

I made a note about not copying build output artifacts back to Windows. This is a good suggestion. At minimum we have to copy some object files for debugging to work, since gdb needs them, if you're debugging in gdbserver mode. But it seems reasonable to give the user the power to override the copy and not having the build system copy the build files. Note that incremental build addresses some of this already here.

One major new feature that we added in 1.0.5, that will be beneficial here, is Makefiles support. With templates as well. With it, you can invoke any external build system, make, scons, CMake etc. The only thing you need to supply (beside how to invoke your build system on the remote system) is the output files that it generates. As soon as you do that, you can do debugging. If you want IntelliSense, you get that to, you can decide what files you want IntelliSense for, and can customize it outside of the build, without affecting the build. With Makefiles you have complete control over your builds. This works well in conjunction with a source copy management feature that we also added in 1.0.5, where you can control, what files gets copy to the remote system and where. And you can do it on a per project or per file basis. With those two you get complete control over build. For example, with the source management feature, you can add your files to the project from a samba share, and specifically tell our Makefile build not to copy them anywhere. You also have the local and remote build events, so you can do lots of interesting things. 

Regarding the generating make files, I think our build doing that is an excellent idea, if anything it gives you more flexibility and reusability, and it's just a cool thing to do. In fact I saw that somewhere else, not sure where, and made a note of it a while ago. It might been you telling me about it on the extension page. The cool thing about our build is that it's really modular and very easy to customize, so it would actually be pretty easy to make it generate make files. We already generate complete remote command lines.

But with all those new features and improvements mentioned above, it's becomes less necessary to generator make or CMake files. It's still a cool thing to do, but lower in priority, we'll still probably do it :)

Thanks
-Ion

**Comment by robotdad** [comment]
@stanthomas should we leave this issue open for further discussion or close it since we have makefile templates in 1.0.5?

**Comment by stanthomas** [comment]
Close because changes in 1.0.5 have moved the discussion on:
1) makefile projects are supported giving the flexibility to take over the build process.
2) incremental builds are supported with limitations discussed in #29
